server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -191,9 +191,7 @@ public static boolean searchFor(RoleModel role, RoleModel composite, Set<String>
             return false;
         }
 
-        Set<RoleModel> compositeRoles = composite.getCompositesStream().collect(Collectors.toSet());
-        return compositeRoles.contains(role) ||
-                        compositeRoles.stream().anyMatch(x -> x.isComposite() && searchFor(role, x, visited));
```

Comment:
Running profiler and looking thread dumps, this one has shown that `collect` is too expensive here. By changing it we save 1s one searching for roles.